packages/core-data/src/utils/block-selection-history.ts

```diff
@@ -0,0 +1,210 @@
+		const block = findBlockByClientIdInDoc( clientId, this.ydoc );
+
+		const attributes = block?.get( 'attributes' ) as
+			| Y.Map< Y.Text >
+			| undefined;
+
+		const attributeKey = selection.attributeKey;
+
+		const changedYText = attributeKey
+			? attributes?.get( attributeKey )
+			: undefined;
+
+		const isYText = changedYText instanceof Y.Text;
+		const isFullyDefinedPosition = attributeKey && clientId;
+
+		if ( ! isYText || ! isFullyDefinedPosition ) {
+			// We either don't have a valid YText (it's been deleted) or we've
+			// been passed a selection that's just a block clientId.
+			// Store as BlockSelection.
+			return {
+				type: YSelectionType.BlockSelection,
+				clientId,
+			};
+		}
+
+		const offset = selection.offset ?? 0;
+		const relativePosition = Y.createRelativePositionFromTypeIndex(
+			changedYText,
+			offset
+		);
+
+		return {
+			type: YSelectionType.RelativeSelection,
+			attributeKey,
+			relativePosition,
+			clientId,
+			offset,
+		};
+	}
+
+	/**
+	 * Update the selection history with a new selection.
+	 * @param newSelection
+	 */
+	public updateSelection( newSelection: WPSelection ): void {
+		if (
+			! newSelection ||
+			! newSelection.selectionStart?.clientId ||
+			! newSelection.selectionEnd?.clientId
+		) {
+			return;
+		}
+
+		const { selectionStart, selectionEnd } = newSelection;
+		const start = this.convertSelectionToPositionPart( selectionStart );
+		const end = this.convertSelectionToPositionPart( selectionEnd );
+
+		const ySelection: YFullSelection = {
+			start,
+			end,
+		};
+
+		// Check if the new selection has the same start and end block combination as current
+		const isSameBlockCombinationAsCurrent =
+			this.currentSelection &&
+			start.clientId === this.currentSelection.start.clientId &&
+			end.clientId === this.currentSelection.end.clientId;
+
+		if ( this.currentSelection && ! isSameBlockCombinationAsCurrent ) {
+			// Only add to history if we're moving to a different block combination
+			this.addToHistory( this.currentSelection );
+		}
+
+		// Remove the new selection from history if it exists there (since it's now current)
+		this.history = this.history.filter( ( entry ) => {
+			const isSameBlockCombination =
+				entry.start.clientId === start.clientId &&
+				entry.end.clientId === end.clientId;
+
+			return ! isSameBlockCombination;
+		} );
+
+		this.currentSelection = ySelection;
+	}
+
+	/**
+	 * Get the current position (most recent selection in the current block).
+	 */
+	public getCurrentSelection(): YFullSelection | null {
+		return this.currentSelection;
+	}
+
+	/**
+	 * Get the block history (previous blocks only, not current or last selection).
+	 * @param count Number of positions to retrieve
+	 */
+	public getSelectionHistory( count: number ): YFullSelection[] {
+		return this.history.slice( 0, count );
+	}
+
+	/**
+	 * Add a position to the history, maintaining only the last `historySize` unique selections.
+	 * New positions are added to the front.
```

Comment:
These are selections, not positions, right? Or am I missing some nuance?
```suggestion
	 * Add a selection to the history, maintaining only the last `historySize` unique selections.
	 * New selections are added to the front.
```

Reply:
That's correct, these should all be called "selections". Some back-and-forth on naming during development. I'll do a find for "position" used incorrectly here and make sure they reflect the correct name.